ducktape/tests/runner.py

```diff
@@ -64,95 +64,120 @@ def __init__(self, cluster, session_context, session_logger, tests, port=Console
         self.active_tests = {}
         self.finished_tests = {}
 
-        self.current_test_context = None
-
     def who_am_i(self):
         """Human-readable name helpful for logging."""
         return self.__class__.__name__
 
     def run_all_tests(self):
         self.results.start_time = time.time()
-        self.log(logging.INFO, "starting test run with session id %s..." % self.session_context.session_id)
-        self.log(logging.INFO, "running %d tests..." % len(self.staged_tests))
+        self._log(logging.INFO, "starting test run with session id %s..." % self.session_context.session_id)
+        self._log(logging.INFO, "running %d tests..." % len(self.staged_tests))
 
         while len(self.staged_tests) > 0 or len(self.active_tests) > 0:
 
             if len(self.active_tests) == 0:
-                self.current_test_context = self.staged_tests.pop()
-                self.run_single_test()
-
+                next_test_context = self.staged_tests.pop()
+                self._preallocate_subcluster(next_test_context)
+                self._run_single_test(next_test_context)
             try:
                 event = self.receiver.recv()
-                self.handle(event)
+                self._handle(event)
             except Exception as e:
                 err_str = "Exception receiving message: %s: %s" % (str(type(e)), str(e))
                 err_str += "\n" + traceback.format_exc(limit=16)
-                self.log(logging.ERROR, err_str)
+                self._log(logging.ERROR, err_str)
                 continue
 
         for proc in self.proc_list:
             proc.join()
 
         return self.results
 
-    def run_single_test(self):
+    def _run_single_test(self, test_context):
         """Start a test runner client in a subprocess"""
         proc = multiprocessing.Process(
             target=run_client,
             args=[
                 self.hostname,
                 self.port,
-                self.current_test_context.test_id,
-                self.current_test_context.logger_name,
-                self.current_test_context.results_dir,
+                test_context.test_id,
+                test_context.logger_name,
+                test_context.results_dir,
                 self.session_context.debug
             ])
         self.proc_list.append(proc)
         proc.start()
 
-    def handle(self, event):
-        self.log(logging.DEBUG, str(event))
+    def _preallocate_subcluster(self, test_context):
+        """Preallocate the subcluster which will be used to run the test.
+
+        Side effect: store association between the test_id and the preallocated subcluster.
+
+        :param test_context
+        :return None
+        """
+
+        expected_num_nodes = test_context.expected_num_nodes
+        if expected_num_nodes is None:
+            # If there is no information on cluster usage, allocate entire cluster
+            expected_num_nodes = len(self.cluster)
```

Comment:
log something? If we're using this runner, at least if it has parallelism > 1, then we probably want to tell the user they need to annotate their tests to get parallelism